### PR TITLE
Update jwt and main functions

### PR DESCRIPTION
**Pull request by abidkhan03**

### CODE REVIEW

File: jwt/jwt.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+type User struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+func init() {
+	TokenAuth = &jwtauth.JWTAuth{}
+}
+
+func signUp(w http.ResponseWriter, r *http.Request) {
```
Comment (IsrarAliKhan) [inline jwt/jwt.go]: Move the whole code in this function to `handlers/auth.go` SignIn function

File: jwt/jwt.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	// Create the JWT string
+	tokenString, err := token.SignedString([]byte("secret"))
+    if err != nil {
+        http.Error(w, err.Error(), http.StatusBadRequest)
+        return
+    }
+    w.Header().Set("Content-Type", "application/json")
+    w.WriteHeader(http.StatusOK)
+    w.Write([]byte(tokenString))
```
Comment (IsrarAliKhan) [inline jwt/jwt.go]: `tokenString` is a secret token. It should not be accessible to anyone. If you send it in the response, anyone can see it and can access your private data.

Tell me what is the best way to send `tokenString`, so no one can access it. Verbally :)

File: jwt/jwt.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+	claims := &abc.MapClaims{
+		"username": user.Username,
+		"exec":     time.Now().Add(time.Hour),
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	// Create the JWT string
+	tokenString, err := token.SignedString([]byte("secret"))
```
Comment (IsrarAliKhan) [inline jwt/jwt.go]:
```suggestion
	tokenString, err := token.SignedString(jwtKey)
```

File: jwt/jwt.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	abc "github.com/dgrijalva/jwt-go"
+	"github.com/go-chi/jwtauth/v5"
+)
+
+var TokenAuth *jwtauth.JWTAuth
```
Comment (IsrarAliKhan) [inline jwt/jwt.go]: Do we need this anymore?

File: jwt/jwt.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+type User struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+func init() {
+	TokenAuth = &jwtauth.JWTAuth{}
+}
```
Comment (IsrarAliKhan) [inline jwt/jwt.go]: Don't you think we should remove this now?

File: jwt/jwt.go (hunk 3 above)
Comment (IsrarAliKhan) [reply jwt/jwt.go]: declare `jwtKey` out side this function i.e.
`var jwtKey = []byte("my_secret_key")`

File: handler/auth.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	username := r.FormValue("username")
+	password := r.FormValue("password")
+
+	// Check the user's credentials
+	if username != "admin" || password != "admin" {
+		http.Error(w, "Unauthorized", 401)
+		return
+	}
+
+
```
Comment (IsrarAliKhan) [inline handler/auth.go]: Paste the correct sign in code here

File: auth/jwt.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		if claims["username"] == "" {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, `{"error": "invalid credentials"}`)
+			return
+		}
+		if claims["username"] == "Ali" {
+			next.ServeHTTP(w, r)
+			return
+		}
```
Comment (IsrarAliKhan) [inline auth/jwt.go:41]: no need to return

File: auth/jwt.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, claims, err := jwtauth.FromContext(r.Context())
+		if err != nil {
+			next.ServeHTTP(w, r)
+			return
+		}
+		
+		if claims["username"] == "" {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
```
Comment (IsrarAliKhan) [inline auth/jwt.go]: It should be unauthorized error

File: auth/jwt.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(tokenString))
+}
+
+func ValidateToken(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, claims, err := jwtauth.FromContext(r.Context())
+		if err != nil {
+			next.ServeHTTP(w, r)
```
Comment (IsrarAliKhan) [inline auth/jwt.go]: If there is an error, it shouldn't be moved to next request

File: auth/jwt.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+		}
+		
+		if claims["username"] == "" {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, `{"error": "invalid credentials"}`)
+			return
+		}
+		if claims["username"] == "Ali" {
+			next.ServeHTTP(w, r)
```
Comment (IsrarAliKhan) [inline auth/jwt.go:40]: Check the expiry date before moving forward. If it is expired, it should not go further to the next

File: auth/jwt.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	// Create the JWT string
+	tokenString, err := token.SignedString([]byte("secret"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(tokenString))
+}
```
Comment (IsrarAliKhan) [inline auth/jwt.go]: This is a `SignIn` handler. It should be in another directory i.e. `handlers/auth.go`

File: auth/jwt.go [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+var TokenAuth *jwtauth.JWTAuth
+
+type User struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+func init() {
+	TokenAuth = jwtauth.New("HS256", []byte("secret"), nil)
```
Comment (IsrarAliKhan) [inline auth/jwt.go:17]: we are using `[]byte("secret")` in multiple places in this file. Declare this variable at one place and use it all other places

**Review by IsrarAliKhan — CHANGES_REQUESTED** [review]

File: auth/jwt.go (hunk 12 above)
Comment (abidkhan03) [reply auth/jwt.go:17]: Fixed

**Pull request closed without merging**